docs/design/copy-tracking.md

```diff
@@ -10,17 +10,613 @@ results of rebasing across remote copies.
+file's past names. For example, if file `foo` gets renamed to `bar` in one
+commit and then to `baz` in another commit, we will record that `baz` previously
+had names `bar` and `foo`.
+
+To support merging two files into one, the list of past names is actually a DAG.
+Merging can happen in a merge commit when two sides copy/rename different
+source files to the same target file. By having support for it in the model, we
+can also support merging multiple files into one in a regular non-merge commit.
+
+To avoid having to store all past paths in the tree object entry, we will write
+the copy history as an object and the tree will refer to the object by ID.
+
+Each node in the graph stores the path. Having the path in the copy graph can be
+useful for finding copy sources without having to scan the whole tree or having
+to ask the backend.
+
+If we use the file name as only input to the ID, then we get deterministic tree
+ids. On the other hand, if we add a salt to the copy graph node, then we can
+represent that a file was rewritten from scratch. That makes logical sense, but
+I'm not sure how useful it will be.
+
+Do we want to allow an edge in the copy graph to be able to say that changes
+should not be propagated across it? It might be useful to still have such edges
+for the log and annotation use cases.
+
+Should we support copy tracking for symlinks? Their history is not very for
+annotation purposes, but knowing the history may at least be useful for
+detecting directory renames (if all files and symlinks in a directory were
+renamed).
+
+We won't support tracking copied directories.
+
+The data structure might look like this:
+```rust
+// Current `TreeValue::File` variant:
+File { id: FileId, executable: bool },
+// New `TreeValue::File` variant:
+File { id: FileId, executable: bool, copy_id: CopyId },
+
+// A CopyId is a hash of this struct:
+struct CopyHistory {
+    path: RepoPath,
+    parents: Vec<CopyId>
+}
+```
+
+### Diffing
+
+When diffing two trees, we first diff the trees without considering copy info.
+For any copy IDs that changed in that diff, we walk all of their copy graphs to
+figure out how they're related and which source file to associate with which
+destination file.
+
+The details of the algorithm is left for the implementation. The following
+sections provide some examples to hopefully show that it's feasible.
+
+#### Example: Divergent copy and rename
+
+Let's look at an example of how this model would look in this scenario:
+
+```
+M rename foo->baz, create bar
+|
+| L copy foo->bar, create baz
+|/
+K add foo
+```
+
+Assuming the new files are different in each commit, we get these trees:
+```
+Commit K:
+name: foo, id: K, copy_id: 1:foo
+
+Commit L:
+name: bar, id: L, copy_id: 2:bar->1:foo
+name: baz, id: L, copy_id: 3:baz
+name: foo, id: K, copy_id: 1:foo
+
+Commit M:
+name: bar, id: M, copy_id: 4:bar
+name: baz, id: M, copy_id: 5:baz->1:foo
```

Comment:
These arrows could be considered "backwards" by analogy to commits. `baz` has a pointer to `foo`, but this represents `foo` becoming `baz`. In the descriptions in the beginning of the section, the arrows go the other way.

My guess is that it's best to point the arrows the other way from the beginning (so, this arrow should be `foo`->`baz`, and the text should say that `baz` has a pointer to `foo`), but it's of course arguable.